core/src/main/java/com/datastrato/gravitino/storage/relational/service/RoleMetaService.java

```diff
@@ -76,16 +92,38 @@ public void insertRole(RoleEntity roleEntity, boolean overwritten) {
           MetalakeMetaService.getInstance().getMetalakeIdByName(roleEntity.namespace().level(0));
       RolePO.Builder builder = RolePO.builder().withMetalakeId(metalakeId);
       RolePO rolePO = POConverters.initializeRolePOWithVersion(roleEntity, builder);
+      List<SecurableObjectPO> securableObjectPOs = Lists.newArrayList();
+      for (SecurableObject object : roleEntity.securableObjects()) {
+        SecurableObjectPO.Builder objectBuilder =
+            POConverters.initializeSecurablePOBuilderWithVersion(
+                roleEntity.id(), object, getEntityType(object));
+        objectBuilder.withEntityId(
+            getSecurableObjectEntityId(metalakeId, object.fullName(), object.type()));
+        securableObjectPOs.add(objectBuilder.build());
+      }
 
-      SessionUtils.doWithCommit(
-          RoleMetaMapper.class,
-          mapper -> {
-            if (overwritten) {
-              mapper.insertRoleMetaOnDuplicateKeyUpdate(rolePO);
-            } else {
-              mapper.insertRoleMeta(rolePO);
-            }
-          });
+      SessionUtils.doMultipleWithCommit(
+          () ->
+              SessionUtils.doWithoutCommit(
+                  SecurableObjectMapper.class,
+                  mapper -> {
+                    if (overwritten) {
+                      mapper.softDeleteSecurableObjectsByRoleId(rolePO.getRoleId());
+                    }
+                    if (!securableObjectPOs.isEmpty()) {
+                      mapper.batchInsertSecurableObjects(securableObjectPOs);
+                    }
+                  }),
+          () ->
+              SessionUtils.doWithoutCommit(
+                  RoleMetaMapper.class,
+                  mapper -> {
+                    if (overwritten) {
+                      mapper.insertRoleMetaOnDuplicateKeyUpdate(rolePO);
```

Comment:
Why do you still need to call insert on duplicate since they're already marked as deleted above?

Reply:
We just deleted the securable objects instead of role meta.